### PR TITLE
UCP/CORE: Set parent memh field for memh created with ALLOCATE flag

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Set parent memh field for memh created with `UCP_MEM_MAP_ALLOCATE` flag.

## Why ?

Fixes the following bug:
```
[ RUN      ] shm_ib_ipc/test_ucp_mmap.fixed/3 <shm,ib,cuda_ipc,rocm_ipc,cuda_copy,rocm_copy/no_rcache>
[New Thread 0x7ffff2540700 (LWP 192078)]
[jazz11:191967:0:191967]    ucp_mm.inl:89   Assertion `context->rcache != ((void *)0)' failed

/hpc/mtr_scrap/users/dmitrygla/ucx/src/ucp/core/ucp_mm.inl: [ ucp_memh_put() ]
      ...
       86         return;
       87     }
       88
==>    89     ucs_assert(context->rcache != NULL);
       90
       91     UCP_THREAD_CS_ENTER(&context->mt_lock);
       92     ucs_rcache_region_put_unsafe(context->rcache, &memh->super);

==== backtrace (tid: 191967) ====
 0 0x000000000006debb ucp_memh_put()  /hpc/mtr_scrap/users/dmitrygla/ucx/src/ucp/core/ucp_mm.inl:89
 1 0x000000000006debb ucp_mem_unmap()  /hpc/mtr_scrap/users/dmitrygla/ucx/src/ucp/core/ucp_mm.c:738
 2 0x00000000008b2e92 test_ucp_mmap_fixed_Test::test_body()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/ucp/test_ucp_mmap.cc:639
 3 0x00000000005a0ccc ucs::test_base::run()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/common/test.cc:378
 4 0x00000000005a0e8f ucs::test_base::TestBodyProxy()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/common/test.cc:404
 5 0x00000000007f26aa ucp_test::TestBody()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/ucp/ucp_test.h:202
 6 0x0000000000e7773a testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/common/googletest/gtest.cc:2433
 7 0x0000000000e735c4 testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/common/googletest/gtest.cc:2469
 8 0x0000000000e5e357 testing::Test::Run()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/common/googletest/gtest.cc:2509
 9 0x0000000000e5ebbf testing::TestInfo::Run()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/common/googletest/gtest.cc:2687
10 0x0000000000e5f278 testing::TestSuite::Run()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/common/googletest/gtest.cc:2819
11 0x0000000000e6a476 testing::internal::UnitTestImpl::RunAllTests()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/common/googletest/gtest.cc:5350
12 0x0000000000e7843d testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/common/googletest/gtest.cc:2433
13 0x0000000000e74468 testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/common/googletest/gtest.cc:2469
14 0x0000000000e68f42 testing::UnitTest::Run()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/common/googletest/gtest.cc:4940
15 0x000000000058749f RUN_ALL_TESTS()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/common/googletest/gtest.h:2473
16 0x000000000058739b main()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/common/main.cc:106
17 0x00000000000223d5 __libc_start_main()  ???:0
18 0x0000000000586b19 _start()  ???:0
=================================
```

## How ?

1. Accept allocation `method` into `ucp_memh_create` function.
2. Rename `ucp_memh_init_from_mem` to `ucp_memh_create_from_mem` and allocate memory for `UCP_MEM_MAP_ALLOCATE`'d memory handle using `ucp_memh_create` function.
3. Add `ucs_assert(memh->parent != NULL)` assertion in `ucp_mem_map` function to make sure all allocated and returned to a user memory handles have `parent` field set.

### CODE REVIEW

File: src/ucp/core/ucp_mm.c [hunk 1]
```diff
@@ -445,17 +445,22 @@ ucp_memh_rcache_get(ucp_context_h context, void *address, size_t length,
     return UCS_OK;
 }
 
-static void
-ucp_memh_init_from_mem(ucp_context_h context, ucp_mem_h memh,
-                       uct_allocated_memory_t *mem)
+static ucs_status_t
+ucp_memh_create_from_mem(ucp_context_h context, uct_allocated_memory_t *mem,
```
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_mm.c:449]: maybe `uct_alloc_method_t` or `const uct_allocated_memory_t`?

**Comment by brminich** [comment]
is it fails with #8539?

**Comment by dmitrygx** [comment]
> does it fail with #8539?

I think so, because now we disable RCACHE as it is expected. The issue reproduced only on jazz

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@Artemy-Mellanox can you pls review?

**Review by brminich — APPROVED** [review]

**Review by Artemy-Mellanox — APPROVED** [review]

**Pull request merged**